File: python/백준/스택, 큐, 덱/12789 도키도키 간식드리미.py
n = int(input())

student = list(map(int, input().split()))
stack = []
count = 1

for i in range(n):
  if student[i] == count:
    count += 1
  else:
    stack.append(student[i])
  
  while stack and stack[-1] == count:
      stack.pop()
      count += 1
      

if stack:
  print("Sad")
else:
  print("Nice")

File: cpp/백준/자료구조/스택, 집합/12789 도키도키 간식드리미.cpp
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<int> student(n);
    for (int i = 0; i < n; ++i) cin >> student[i];

    stack<int> st;
    int count = 1;

    for (int i = 0; i < n; ++i) {
        if (student[i] == count) {
            ++count;
        } else {
            st.push(student[i]);
        }

        while (!st.empty() && st.top() == count) {
            st.pop();
            ++count;
        }
    }

    cout << (st.empty() ? "Nice" : "Sad") << "\n";
    return 0;
}
